raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java

```diff
@@ -88,56 +94,90 @@ public void punctuate(long timestamp) {
           HexUtils.getHex(key.getTraceId()));
       return;
     }
 
+    long emitTs = traceState.getEmitTs();
     if (emitTs <= timestamp) {
       // we can emit this trace so just delete the entry for this 'key'
-      traceEmitTriggerStore.delete(key);
       // Implies that no new spans for the trace have arrived within the last
       // 'groupingWindowTimeoutMs' interval
       // so the trace can be finalized and emitted
-      ValueAndTimestamp<RawSpans> agg = inflightTraceStore.delete(key);
-      ByteBuffer traceId;
-      String tenantId;
-      if (agg != null && !agg.value().getRawSpans().isEmpty()) {
-        RawSpan firstSpan = agg.value().getRawSpans().get(0);
-        traceId = firstSpan.getTraceId();
-        tenantId = firstSpan.getCustomerId();
-
-        List<RawSpan> rawSpanList = agg.value().getRawSpans();
-        recordSpansPerTrace(rawSpanList.size(), List.of(Tag.of("tenant_id", tenantId)));
-        Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
-        StructuredTrace trace =
-            StructuredTraceBuilder.buildStructuredTraceFromRawSpans(
-                rawSpanList, traceId, tenantId, timestamps);
+      traceEmitTriggerStore.delete(key);
 
+      ByteBuffer traceId = traceState.getTraceId();
+      String tenantId = traceState.getTenantId();
+
+      TreeSet<ByteBuffer> spanIdsSet =
+          new TreeSet<>(
+              (spanId1, spanId2) -> {
+                // the sort order for this set is determined by comparing the corresponding {@code
+                // SpanIdentifier}
+                Serde<SpanIdentity> serde = (Serde<SpanIdentity>) context.keySerde();
+                byte[] spanIdentityBytes1 =
+                    serde.serializer().serialize("", new SpanIdentity(tenantId, traceId, spanId1));
+                byte[] spanIdentityBytes2 =
+                    serde.serializer().serialize("", new SpanIdentity(tenantId, traceId, spanId2));
+                return ByteBuffer.wrap(spanIdentityBytes1)
+                    .compareTo(ByteBuffer.wrap(spanIdentityBytes2));
+              });
+      spanIdsSet.addAll(traceState.getSpanIds());
+
+      List<RawSpan> rawSpanList = new ArrayList<>();
+      try (KeyValueIterator<Windowed<SpanIdentity>, RawSpan> iterator =
+          spanWindowStore.fetch(
+              new SpanIdentity(tenantId, traceId, spanIdsSet.first()),
+              new SpanIdentity(tenantId, traceId, spanIdsSet.last()),
+              Instant.ofEpochMilli(traceState.getTraceStartTimestamp()),
+              Instant.ofEpochMilli(traceState.getTraceEndTimestamp()))) {
+        iterator.forEachRemaining(
+            keyValue -> {
+              // range search could return extra span ids as well, filter them
+              // one scenario when this could occur is when some spanIdentifier for another trace
+              // has byte value which is within the {@code spanIdsSet.first()} & {@code
+              // spanIdsSet.last()}
```

Comment:
spans belonging to a different trace shouldn't be in the spanIdsSet - is the comment correct ?

Reply:
yes it is correct